In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import SGDRegressor
import seaborn as sns
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import GridSearchCV

In [3]:
# Download and load the data
import keras
import os

f_path_1 = "data/Euro_Salary.csv"
url_1 = "https://github.com/AkeemSemper/ml_data/raw/main/Euro_Salary.csv"
if not os.path.exists(f_path_1):
    file_1 = keras.utils.get_file(f_path_1, url_1)
df = pd.read_csv(f_path_1)

/Users/ghunaym/Library/Python/3.9/lib/python/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


# Assignment 2 - Regression
### Predict the TOTAL COMPENSATION for this year. 

The data file provided is a salary survey for tech workers in Europe. We want to predict the total amount of compensation they bring in each year, based off of the details of their work. 

Some notes that will be important:
<ul>
    <li>The total compensation will need to be constructed, there is a column for salary, "Yearly brutto salary (without bonus and stocks) in EUR", as well as a column for bonus compensation, "Yearly bonus + stocks in EUR". 
    <li>Some categorical variables will need some work, and there isn't generally an exact answer. The main concern is things with categories that have a bunch of values with a very small count. For example, if there is only 1 person in City X, then that value likely needs to be addressed. We don't want it encoded into a new column of one 1 and thousands of 0s. 
    <li>There is an article exploring some of the data here: https://www.asdcode.de/2021/01/it-salary-survey-december-2020.html
    <li>Imputation and a bit of data manipulation will be required. 
    <li>Use any regression method you'd like. Some ones are closely related to what we've done, you may want to look at them, e.g. ExtraTreesRegressor. 
    <li>Initial accurracy, and potentially final accuracy, may not be great. When I made a plain model will little optimization the errors were large and the R2 was low. There is lots of room for optimization. 
    <li>Research challenge - try some work on the target, look into TransformedTargetRegressor and see if that helps. Recall in stats when we had skewed distributions... Maybe it helps, maybe it doesn't. 
    <li>EDA and data prep are up to you - you'll probably need to do a little exploring to figure out what cleanup is needed. When I did it, I did things kind of iteratively when I did it. For example, look at the value counts, figure out how to treat the different categories, clean something up, look at the results, potentially repeat if needed. After you figure out what needs to be done, you may be able to take some of those steps and incorporate them into a pipeline to be cleaner....
    <li><b>CRITICAL - Please make sure you publish it after having run it, all the output should be showing.</b>
</ul>

### Details and Deliverables

You'll need to build code to produce the predictions. In particular, there's a few things that'll be marked:
<ul>
    <li>Please add a "presentation version" at the bottom, where you show what you did, and the results. Basically, you start with the original data, you do some work to figure out what's needed, you try a few models and select the best. At the bottom, put what <i>you actually settled on</i>, i.e. after all the figuring and exploring, here's the code that goes from raw data to final results, and here's what the results were. I should be able to read this part and understand what you did clearly:
    <ul>
        <li> Please make a pipeline that does the prep work - you may need some exploration or several trials before settling on what exactly to use, that's normal. Once you've settled, build that into a pipeline so it's clear and repeatable.
        <li> What you settled on for data cleaning, along with what prompted it. 
        <li> Feature Selection - Please identify what you did for feature selection. No need for a long explaination, something along the lines of "I did X, and the result was that 4 features were removed". Try at least 2 things. 
        <li> Model selection - between selecting a model style and tuning it with hyperparameters, what did you test and what won?
        <li> Overall, how good was your model and what things may make sense to try to do even better? 
        <li> If you could use titles/bullet points I'd really appreciate it. 
    </ul>
    <li>Grade Breakdown:
    <ul>
        <li> Code is readable, there are comments: 20%
        <li> Explaination as defined above: 60% (20% each point)
        <li> Accuracy: 20% As compared to everyone else. This will be generously graded, I won't be surprised if overall accuracy is low for most people. 
    </ul>
</ul>

<b>The biggest challenge here is translating the data into something useful and clean. This will probably require a bit of exploration, examining the data, thinking about what it means, trying something, then making a model to see what the results are. In particular, think about what value some of the less clean bits of data my hold - binning/grouping, numerical transformations, outlier removal, etc... are all likely to be useful somewhere. You almost certainly need to look at it column by column and make a decision, I'll apologize up front, it isn't the most fun process in the world. There is not one specific correct answer.</b>

In [4]:
#Load Data
#df = pd.read_csv("data/Euro_Salary.csv")
df.describe(include="all").T

count unique  \
Timestamp                                             1253   1248   
Age                                                 1226.0    NaN   
Gender                                                1243      3   
City                                                  1253    119   
Position                                              1247    148   
Total years of experience                             1237     48   
Years of experience in Germany                        1221     53   
Seniority level                                       1241     24   
Your main technology / programming language           1126    256   
Other technologies/programming languages you us...    1096    562   
Yearly brutto salary (without bonus and stocks)...  1253.0    NaN   
Yearly bonus + stocks in EUR                           829    168   
Number of vacation days                               1185     45   
Employment status                                     1236     11   
Сontract duration                                     1224      3   
Main language at work                                 1237     14   
Company size                                          1235      5   
Company type                                          1228     63   

                                                                        top  \
Timestamp                                               24/11/2020 13:55:19   
Age                                                                     NaN   
Gender                                                                 Male   
City                                                                 Berlin   
Position                                                  Software Engineer   
Total years of experience                                                10   
Years of experience in Germany                                            2   
Seniority level                                                      Senior   
Your main technology / programming language                            Java   
Other technologies/programming languages you us...  Javascript / Typescript   
Yearly brutto salary (without bonus and stocks)...                      NaN   
Yearly bonus + stocks in EUR                                              0   
Number of vacation days                                                  30   
Employment status                                        Full-time employee   
Сontract duration                                        Unlimited contract   
Main language at work                                               English   
Company size                                                          1000+   
Company type                                                        Product   

                                                    freq             mean  \
Timestamp                                              2              NaN   
Age                                                  NaN        32.509788   
Gender                                              1049              NaN   
City                                                 681              NaN   
Position                                             387              NaN   
Total years of experience                            138              NaN   
Years of experience in Germany                       195              NaN   
Seniority level                                      565              NaN   
Your main technology / programming language          184              NaN   
Other technologies/programming languages you us...    44              NaN   
Yearly brutto salary (without bonus and stocks)...   NaN  80279042.578715   
Yearly bonus + stocks in EUR                         227              NaN   
Number of vacation days                              488              NaN   
Employment status                                   1190              NaN   
Сontract duration                                   1159              NaN   
Main la

In [5]:
df[393:396]

Timestamp   Age Gender    City          Position   \
393  24/11/2020 19:15:02  30.0   Male  Moscow  Software Engineer   
394  24/11/2020 19:15:49  33.0   Male  Berlin    Product Manager   
395  24/11/2020 19:19:30  35.0   Male  Berlin        QA Engineer   

    Total years of experience Years of experience in Germany Seniority level  \
393                         5                              0          Middle   
394                         5                              5          Senior   
395                        11                             10          Senior   

    Your main technology / programming language  \
393                                           C   
394                                         NaN   
395                                        Java   

    Other technologies/programming languages you use often  \
393                                              C/C++       
394                                             Python       
395  Python, Javascript / Typescript, .NET, Java / ...       

     Yearly brutto salary (without bonus and stocks) in EUR  \
393                                            14712.0        
394                                            70000.0        
395                                            74400.0        

    Yearly bonus + stocks in EUR Number of vacation days   Employment status  \
393                            0                      30  Full-time employee   
394                          800                      30  Full-time employee   
395                          NaN                      30  Full-time employee   

      Сontract duration Main language at work Company size Company type  
393  Unlimited contract               Russian     101-1000      Product  
394                 NaN                German     101-1000      Product  
395  Unlimited contract               English     101-1000      Product

In [6]:
#df.info()
df["Other technologies/programming languages you use often"].value_counts()

Other technologies/programming languages you use often
Javascript / Typescript                                                                   44
Python                                                                                    37
SQL                                                                                       31
AWS, Docker                                                                               16
Kotlin                                                                                    15
                                                                                          ..
Python, Ruby, Java / Scala, Go, Rust, AWS, Kubernetes, Docker, Terraform, helm, packer     1
Kotlin, SQL, AWS, Kubernetes, Docker                                                       1
Python, R, Go                                                                              1
C/C++, PHP, SQL                                                                            1
Python, AWS, Ku

In [7]:
df.sample(20)

Timestamp   Age Gender                   City  \
764   26/11/2020 11:50:35  31.0   Male                 Berlin   
804   26/11/2020 22:41:33  20.0   Male                 Berlin   
454   24/11/2020 22:22:57  33.0   Male                 Berlin   
1146  11/12/2020 16:25:59  28.0   Male                 Berlin   
1048  02/12/2020 00:27:03  31.0   Male                 Berlin   
793   26/11/2020 21:13:28  32.0   Male                 Berlin   
1222  05/01/2021 22:56:24  34.0   Male            Ingolstadt    
150   24/11/2020 12:12:35  25.0   Male                 Berlin   
1067  03/12/2020 08:53:41  32.0   Male                 Berlin   
220   24/11/2020 13:32:34  36.0   Male                 Berlin   
1023  01/12/2020 10:04:53  36.0   Male              Stockholm   
640   25/11/2020 16:50:22  49.0   Male                 Munich   
64    24/11/2020 11:38:10  32.0   Male                 Munich   
615   25/11/2020 14:48:38  32.0   Male              Amsterdam   
388   24/11/2020 19:07:37  35.0   Male                Hamburg   
365   24/11/2020 18:31:25   NaN   Male                 Berlin   
344   24/11/2020 17:45:03  26.0   Male              Frankfurt   
1017  01/12/2020 08:49:09  35.0   Male  Murnau am Staffelsee    
336   24/11/2020 17:37:32  32.0   Male                 Berlin   
347   24/11/2020 17:47:42  27.0   Male                 Munich   

                 Position  Total years of experience  \
764            ML Engineer                         7   
804       Mobile Developer                         5   
454      Software Engineer                       NaN   
1146           ML Engineer                         5   
1048     Software Engineer                         7   
793   Senior Data Engineer                         9   
1222     Software Engineer                       3.5   
150      Backend Developer                         4   
1067     Software Engineer                        13   
220    Engineering Manager                         7   
1023      Mobile Developer                        14   
640      Software Engineer                        27   
64      Frontend Developer                         7   
615      Backend Developer                        15   
388      Software Engineer                        13   
365      Software Engineer                         9   
344      Software Engineer                         3   
1017               Beikoch                         1   
336      Backend Developer                         4   
347      Software Engineer                         5   

     Years of experience in Germany Seniority level  \
764                               2          Middle   
804                               2            Lead   
454                              10          Senior   
1146                              3          Middle   
1048                              2          Senior   
793                               4          Senior   
1222                            3.5          Junior   
150                               1          Middle   
1067                              5            Lead   
220                               1            Lead   
1023                              7          Senior   
640                              19          Senior   
64                                5          Senior   
615                               0          Senior   
388                               3            Lead   
365                               1            Lead   
344                               1          Middle   
1017                              1          Middle   
336                               1          Senior   
347                               1          Senior   

     Your main technology / programming language  \
764                                       Python   
804                                           C#   
454                                         Java   
1146                             Python, Pytorch   
1048                                        Java   
793  

In [17]:
import numpy as np
import pandas as pd

base_salary = "Yearly brutto salary (without bonus and stocks) in EUR"
bonus_col = "Yearly bonus + stocks in EUR"

# Convert to numeric safely
df[base_salary] = pd.to_numeric(df[base_salary], errors="coerce")
df[bonus_col] = pd.to_numeric(df[bonus_col], errors="coerce")

# Fill bonus NaNs with 0
df[bonus_col] = df[bonus_col].fillna(0)

# Drop rows missing base salary
df = df.dropna(subset=[base_salary])

# Target
df["TOTAL_COMP"] = df[base_salary] + df[bonus_col]

df[["TOTAL_COMP", base_salary, bonus_col]].describe()

TOTAL_COMP  Yearly brutto salary (without bonus and stocks) in EUR  \
count  1.253000e+03                                       1.253000e+03        
mean   8.428361e+07                                       8.027904e+07        
std    2.829182e+09                                       2.825061e+09        
min    1.000100e+04                                       1.000100e+04        
25%    6.000000e+04                                       5.880000e+04        
50%    7.400000e+04                                       7.000000e+04        
75%    9.100000e+04                                       8.000000e+04        
max    1.000000e+11                                       1.000000e+11        

       Yearly bonus + stocks in EUR  
count                  1.253000e+03  
mean                   4.004569e+06  
std                    1.412516e+08  
min                    0.000000e+00  
25%                    0.000000e+00  
50%                    0.000000e+00  
75%                    9.400000e+03  
max                    5.000000e+09

In [28]:
# Drop rows missing base salary
df = df.dropna(subset=[base_salary])

# Target
df["TOTAL_COMP"] = df[base_salary] + df[bonus_col]

df[["TOTAL_COMP", base_salary, bonus_col]].describe()

# Drop OUtleirs
low, high = df["TOTAL_COMP"].quantile([0.01, 0.99])

df = df[(df["TOTAL_COMP"] >= low) & (df["TOTAL_COMP"] <= high)].copy()

print("Kept range:", low, "to", high)
df["TOTAL_COMP"].describe()

Kept range: 28081.0 to 218650.0000000001


count      1202.000000
mean      81642.002978
std       32244.427740
min       28300.000000
25%       62000.000000
50%       74000.000000
75%       90000.000000
max      215000.000000
Name: TOTAL_COMP, dtype: float64

In [ ]:
def group_rare_categories(df, col, threshold=100, other_label="Other"):
    vc = df[col].astype(str).fillna("Missing").value_counts()
    rare_values = vc[vc < threshold].index
    df[col] = df[col].astype(str).fillna("Missing").replace(rare_values, other_label)
    return df

candidate_cols = ["City", "City ", "Position", "Position ", "Country", "Country of residence", "Company size"]

# Keep only the ones that exist in this dataset
high_card_cols = [c for c in candidate_cols if c in df.columns]

print("High-card cols found:", high_card_cols)

# Group rare categories
for c in high_card_cols:
    df = group_rare_categories(df, c, threshold=100)

# preview
df[high_card_cols].head()

High-card cols found: ['City', 'Position ', 'Company size']


City          Position  Company size
0  Munich  Software Engineer       51-100
1  Berlin  Backend Developer     101-1000
3  Berlin              Other       51-100
4  Berlin  Backend Developer     101-1000
5  Berlin              Other        11-50

In [ ]:
# Columns that might cause feature explosion
cols_to_drop = [
    "Timestamp",        
    "City",             
    "Position "           
]

X = df.drop(columns=[base_salary, bonus_col, "TOTAL_COMP"] + cols_to_drop, errors="ignore")
y = df["TOTAL_COMP"]

X.shape, y.shape

((1202, 13), (1202,))

In [ ]:
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer

# Identify types
numeric_features = X.select_dtypes(include=["int64", "float64"]).columns
categorical_features = X.select_dtypes(include=["object"]).columns

numeric_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

categorical_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(handle_unknown="ignore", min_frequency=10))
])

preprocessor = ColumnTransformer([
    ("num", numeric_pipeline, numeric_features),
    ("cat", categorical_pipeline, categorical_features)
])

In [32]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [33]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

lin_model = Pipeline([
    ("preprocess", preprocessor),
    ("model", LinearRegression())
])

lin_model.fit(X_train, y_train)
lin_preds = lin_model.predict(X_test)

lin_rmse = np.sqrt(mean_squared_error(y_test, lin_preds))
print("Linear Regression RMSE:", lin_rmse)

Linear Regression RMSE: 33936.75005504474


In [34]:
from sklearn.linear_model import Lasso

lasso_model = Pipeline([
    ("preprocess", preprocessor),
    ("model", Lasso(alpha=0.1, max_iter=50000, random_state=42))
])

lasso_model.fit(X_train, y_train)
lasso_preds = lasso_model.predict(X_test)

lasso_rmse = np.sqrt(mean_squared_error(y_test, lasso_preds))
print("Lasso RMSE:", lasso_rmse)

Lasso RMSE: 34033.23299170564


In [35]:
alphas = [0.01, 0.05, 0.1, 0.2, 0.5, 1.0]

for a in alphas:
    model = Pipeline([
        ("preprocess", preprocessor),
        ("model", Lasso(alpha=a, max_iter=50000, random_state=42))
    ])
    scores = cross_val_score(
        model, X_train, y_train,
        scoring="neg_root_mean_squared_error",
        cv=5
    )
    print(f"alpha={a:<4}  CV RMSE={-scores.mean():,.0f}")

/Users/ghunaym/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_coordinate_descent.py:656: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 21708027503.91217, tolerance: 72507784.46578053
  model = cd_fast.sparse_enet_coordinate_descent(
/Users/ghunaym/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_coordinate_descent.py:656: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 4302076491.571106, tolerance: 76734107.99772382
  model = cd_fast.sparse_enet_coordinate_descent(
/Users/ghunaym/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_coordinate_descent.py:656: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 3401831910.5740356, tolerance: 74243671.7183829
  model = cd_fast.sparse_enet_coordinate_descent(


alpha=0.01  CV RMSE=27,585
alpha=0.05  CV RMSE=27,567
alpha=0.1   CV RMSE=27,567
alpha=0.2   CV RMSE=27,523
alpha=0.5   CV RMSE=27,521
alpha=1.0   CV RMSE=27,518


# Answers and Explainations
(Expand/modify as needed)

### Here's the Data Cleaning Steps I Used

- First, I converted the salary and bonus columns to numeric values because some entries were stored as text. I filled missing bonus values with 0 and removed rows where base salary was missing since it is required to calculate total compensation.
- I created a new column called TOTAL_COMP by adding base salary and bonus together.
- When looking at the summary statistics, I noticed some extremely large salary values that were not realistic. To prevent these outliers from affecting the model too much, I removed observations outside the 1st and 99th percentile range.
- For categorical columns like City and Position that had many unique values, I grouped rare categories into an “Other” category to reduce noise and keep the model more stable.

### Here's my Tuning/Feature Selection Steps

- I used Lasso regression to help reduce overfitting and handle the large number of categorical features created from one-hot encoding.
- I tested different alpha values (0.01 to 1.0) using 5-fold cross-validation and compared their RMSE scores. As alpha increased, the model became more regularized. The best cross-validation result occurred at alpha = 1.0.
- Grouping rare categories and removing outliers also helped reduce convergence warnings and improve model stability.

### Here's my Model's Performance

- The Linear Regression model achieved a test RMSE of about 33,937.
- The Lasso model achieved a test RMSE of about 34,033.
- Using cross-validation, the best alpha value (1.0) produced a CV RMSE of approximately 27,518.
- This means that, on average, the model’s predictions are within about 27k–34k of the true salary values.

### Here's my Final Conclusion on What Worked Best
- The biggest improvements came from cleaning the data properly. Removing extreme outliers and grouping rare categories had a noticeable impact on model performance.
- Lasso and Linear Regression performed very similarly, but Lasso with alpha = 1.0 gave slightly better cross-validation results and helped control noisy features.
- Overall, careful preprocessing made more difference than changing the model itself.